src/service_node/http_server.h

```diff
@@ -0,0 +1,32 @@
+#pragma once
+
+#include "net/http_server_impl_base.h"
+
+#include <cstdio>
+
+namespace service_node
+{
+  class Storage;
+
+  using connection_context = epee::net_utils::connection_context_base;
+  using http_request_info = epee::net_utils::http::http_request_info;
+  using http_response_info = epee::net_utils::http::http_response_info;
+
+  class storage_server : public epee::http_server_impl_base<storage_server>
+  {
+   public:
+    storage_server(const std::string& db_path);
+    ~storage_server();
+    bool handle_http_request(const http_request_info &query_info,
+                             http_response_info &response,
+                             connection_context &m_conn_context) override;
+    bool handle_http_request_map(const http_request_info &query_info,
+                                 http_response_info &response_info,
+                                 connection_context &m_conn_context);
+    bool on_store_message(const http_request_info& query_info, http_response_info& response);
+    bool on_retrieve_message(const http_request_info& query_info, http_response_info& response);
+
+   private:
+    std::unique_ptr<Storage> m_storage;
```

Comment:
One reason to use a smart pointer here is to reduce coupling between compilation units, and thus reduce compilation time. (This might have crept in from my original implementation.)

Reply:
I've spent 1/2 day trying to get it to work with m_storage on the stack, but without success.
I think the issue was that all the headers including cryptonote_core.h now also have to have the storage's directory in their include dirs, which is only inherited when you link against the target.
By making this a pointer, we don't have that issue.

Reply:
This patch was what I changed and it seemed to compile fine.

```
diff --git a/external/loki-storage-server b/external/loki-storage-server
--- a/external/loki-storage-server
+++ b/external/loki-storage-server
@@ -1 +1 @@
-Subproject commit bbe6771979ab1d3ccaaa9581cd198ad06c9d3b5f
+Subproject commit bbe6771979ab1d3ccaaa9581cd198ad06c9d3b5f-dirty
diff --git a/src/service_node/http_server.cpp b/src/service_node/http_server.cpp
index 769fe4d2b..fe2ef4037 100644
--- a/src/service_node/http_server.cpp
+++ b/src/service_node/http_server.cpp
@@ -1,5 +1,4 @@
 #include "http_server.h"
-#include "Storage.hpp"
 #include "pow.hpp"
 #include "utils.hpp"
 
@@ -11,12 +10,8 @@
 namespace service_node {
 
   using http_header_info = epee::net_utils::http::http_header_info;
-
-  storage_server::~storage_server() {
-  }
-
   storage_server::storage_server(const std::string& db_path)
-    : m_storage(new Storage(db_path))
+    : m_storage(db_path)
   {}
 
   bool storage_server::handle_http_request_map(const http_request_info& query_info,
@@ -121,7 +116,7 @@ namespace service_node {
 
     bool success = false;
     try {
-      success = m_storage->store(messageHash, recipient, query_info.m_body, ttlInt);
+      success = m_storage.store(messageHash, recipient, query_info.m_body, ttlInt);
     } catch(std::exception e) {
       printf("Caught exception : %s\n", e.what());
     }
@@ -152,7 +147,7 @@ namespace service_node {
     std::vector<storage::Item> items;
 
     try {
-      success = m_storage->retrieve(recipient, items, lastHash);
+      success = m_storage.retrieve(recipient, items, lastHash);
     } catch (std::exception e) {
       printf("Caught exception: %s\n", e.what());
     }
diff --git a/src/service_node/http_server.h b/src/service_node/http_server.h
index f8f510049..4aa9ed7f0 100644
--- a/src/service_node/http_server.h
+++ b/src/service_node/http_server.h
@@ -1,13 +1,12 @@
 #pragma once
 
 #include "net/http_server_impl_base.h"
+#include "loki-storage-server/storage/include/Storage.hpp"
 
 #include <cstdio>
 
 namespace service_node
 {
-  class Storage;
-
   using connection_context = epee::net_utils::connection_context_base;
   using http_request_info = epee::net_utils::http::http_request_info;
   using http_response_info = epee::net_utils::http::http_response_info;
@@ -16,7 +15,6 @@ namespace service_node
   {
    public:
     storage_server(const std::string& db_path);
-    ~storage_server();
     bool handle_http_request(const http_request_info &query_info,
                              http_response_info &response,
                              connection_context &m_conn_context) override;
@@ -27,6 +25,6 @@ namespace service_node
     bool on_retrieve_message(const http_request_info& query_info, http_response_info& response);
 
    private:
-    std::unique_ptr<Storage> m_storage;
+    Storage m_storage;
   };
 } // namespace service_node
```

Reply:
Including the whole path did the trick: "loki-storage-server/storage/include/Storage.hpp".
I was convinced that cmake would add the paths correctly to the include dirs, hence I had only tried with "Storage.hpp".
The idea behind having Storage and DatabaseHandler was to provide an interface that wasn't database specific (e.g. we want to swap sqlite with something else in the future). But I hear you say "YAGNI" so I can remove it.